### PR TITLE
archive/zstd: Fix faulty position check

### LINKED ISSUE #999: archive/zstd instantly blows up when fuzzed
`assert(last_pos > 0)` in zstd.cpp fails when trying to fuzz archive/zstd. It also fails on the commit that adds zstd.
https://github.com/robinlinden/hastur/blob/128ee1c2c9071447d1d2a4092d28f1353b906611/archive/zstd.cpp#L83

Tested with Clang 18 + libstdc++13.

I tried changing it to `assert(last_pos > 0 || last_ret != 0)` since `last_ret != 0` returns `ZstdError::DecodeEarlyTermination`, but that also fails, so not sure what the assert is meant to be checking.

I discovered this when running a fuzz sanity-check on my changes in #998, but I guess they were innocent.

```
robin@w:~/code/hastur2$ git checkout 8f54f7ed1d511986fe73fdf232b5d3a5affb4fa7
HEAD is now at 8f54f7e archive: Add zstd decoding

robin@w:~/code/hastur2$ bazel run archive:zstd_fuzz_test_run --config libfuzzer -c dbg
Starting local Bazel server and connecting to it...
INFO: Analyzed target //archive:zstd_fuzz_test_run (60 packages loaded, 2986 targets configured).
INFO: Found 1 target...
Target //archive:zstd_fuzz_test_run up-to-date:
  bazel-bin/archive/zstd_fuzz_test_run
INFO: Elapsed time: 10.053s, Critical Path: 2.02s
INFO: 53 processes: 19 internal, 34 linux-sandbox.
INFO: Build completed successfully, 53 total actions
INFO: Running command line: bazel-bin/archive/zstd_fuzz_test_run
Launching archive/zstd_fuzz_test_bin as a libFuzzer fuzz test...
Using test output root: /tmp/fuzzing
Writing new corpus elements at: /tmp/fuzzing/corpus
Writing new artifacts at: /tmp/fuzzing/artifacts
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 3273946248
INFO: Loaded 1 modules   (4455 inline 8-bit counters): 4455 [0x563004ae60c0, 0x563004ae7227),
INFO: Loaded 1 PC tables (4455 PCs): 4455 [0x563004ae7228,0x563004af8898),
INFO:     5232 files found in /tmp/fuzzing/corpus
INFO:        0 files found in archive/zstd_fuzz_test_corpus
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 8192 bytes
INFO: seed corpus: files: 5232 min: 1b max: 8192b total: 1798790b rss: 28Mb
zstd_fuzz_test_raw_: archive/zstd.cpp:83: tl::expected<std::vector<std::uint8_t>, ZstdError> archive::zstd_decode(const std::span<const uint8_t>): Assertion `last_pos > 0' failed.
==24988== ERROR: libFuzzer: deadly signal
    #0 0x563004a479a4 in __sanitizer_print_stack_trace (/home/robin/.cache/bazel/_bazel_robin/016f76661bac9ffbafd0e28fd68145ed/execroot/__main__/bazel-out/k8-dbg-ST-64ff5593027f/bin/archive/zstd_fuzz_test_raw_+0x819a4) (BuildId: a6d86014e5c91fb26570b96f5ae4b165693654d1)
    #1 0x563004a1d838 in fuzzer::PrintStackTrace() crtstuff.c
    #2 0x563004a03ed3 in fuzzer::Fuzzer::CrashCallback() crtstuff.c
    #3 0x7f39564f051f  (/lib/x86_64-linux-gnu/libc.so.6+0x4251f) (BuildId: 490fef8403240c91833978d494d39e537409b92e)
    #4 0x7f39565449fb in __pthread_kill_implementation nptl/pthread_kill.c:43:17
    #5 0x7f39565449fb in __pthread_kill_internal nptl/pthread_kill.c:78:10
    #6 0x7f39565449fb in pthread_kill nptl/pthread_kill.c:89:10
    #7 0x7f39564f0475 in gsignal signal/../sysdeps/posix/raise.c:26:13
    #8 0x7f39564d67f2 in abort stdlib/abort.c:79:7
    #9 0x7f39564d671a in __assert_fail_base assert/assert.c:92:3
    #10 0x7f39564e7e95 in __assert_fail assert/assert.c:101:3
    #11 0x563004a48d89 in archive::zstd_decode(std::span<unsigned char const, 18446744073709551615ul>) /proc/self/cwd/archive/zstd.cpp:83:5
    #12 0x563004a47e3d in LLVMFuzzerTestOneInput /proc/self/cwd/archive/zstd_fuzz_test.cpp:14:19
    #13 0x563004a05470 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) crtstuff.c
    #14 0x563004a04be5 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) crtstuff.c
    #15 0x563004a06876 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::vector<fuzzer::SizedFile, std::allocator<fuzzer::SizedFile>>&) crtstuff.c
    #16 0x563004a06de3 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, std::allocator<fuzzer::SizedFile>>&) crtstuff.c
    #17 0x5630049f50df in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) crtstuff.c
    #18 0x563004a1e192 in main (/home/robin/.cache/bazel/_bazel_robin/016f76661bac9ffbafd0e28fd68145ed/execroot/__main__/bazel-out/k8-dbg-ST-64ff5593027f/bin/archive/zstd_fuzz_test_raw_+0x58192) (BuildId: a6d86014e5c91fb26570b96f5ae4b165693654d1)
    #19 0x7f39564d7d8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #20 0x7f39564d7e3f in __libc_start_main csu/../csu/libc-start.c:392:3
    #21 0x5630049ea654 in _start (/home/robin/.cache/bazel/_bazel_robin/016f76661bac9ffbafd0e28fd68145ed/execroot/__main__/bazel-out/k8-dbg-ST-64ff5593027f/bin/archive/zstd_fuzz_test_raw_+0x24654) (BuildId: a6d86014e5c91fb26570b96f5ae4b165693654d1)

NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
MS: 0 ; base unit: 0000000000000000000000000000000000000000
0x28,
(
artifact_prefix='/tmp/fuzzing/artifacts/'; Test unit written to /tmp/fuzzing/artifacts/crash-28ed3a797da3c48c309a4ef792147f3c56cfec40
Base64: KA==
```

### PR DESCRIPTION
**Pull request by Zer0-One**
Closes #999 .

### CODE REVIEW

**Review by robinlinden — COMMENTED** [review]
The fix is probably good, but can you also add a unit test hitting the `last_pos == 0` case?

You can get input for it from the fuzzer if you don't know how to create it normally. It's printed at the end if things fail, so e.g. `Base64: VA==` from one of my runs when setting up the brief fuzzing in CI.

**Comment by Zer0-One** [comment]
There was a bug in the output buffer sizing. It should be fixed now, and the new test cases check for that `last_pos == 0` condition when the output is zero-length, and when it's exactly the chunk size.

File: archive/zstd.cpp [hunk 1]
```diff
@@ -77,16 +78,19 @@ tl::expected<std::vector<std::uint8_t>, ZstdError> zstd_decode(std::span<uint8_t
...
         return tl::unexpected{ZstdError::DecodeEarlyTermination};
     }
 
-    auto const out_size = (chunk_size * count) - (chunk_size * count - last_pos);
+    std::size_t out_size = 0;
+
+    if (count == 1) {
+        out_size = last_pos;
+    } else {
+        out_size = last_pos == 0 ? chunk_size * count : (chunk_size * count) - (chunk_size * count - last_pos);
```
Comment (robinlinden) [inline archive/zstd.cpp:92]: The falsy case in the ternary looks off; you sure it's correct?

`(chunk_size * count) - (chunk_size * count - last_pos)` can be simplified to `last_pos`.

Comment (Zer0-One) [reply archive/zstd.cpp:92]: no, it's wrong. too much copy+paste from testing. fixing it now

Comment (Zer0-One) [reply archive/zstd.cpp:92]: Also gonna add a test for multiples of chunk_size > 1, which would have caught this earlier

Comment (robinlinden) [reply archive/zstd.cpp:92]: Sounds good! I missed it the first review too. Good test cases is hard. :(

Comment (robinlinden) [reply archive/zstd.cpp]: Test case for `count != 1 && last_pos > 0` where this bug was?

Comment (Zer0-One) [reply archive/zstd.cpp]: done

**Review by robinlinden — APPROVED** [review]
Nice job. :P I had no idea of what needed to be done to properly fix the assert, but seeing it now, I guess it required a bunch of work.

**Pull request merged**